test/subdomainregistrar/TestSubdomainRegistrar.js

```diff
@@ -0,0 +1,225 @@
+    account2 = await signers[1].getAddress()
+
+    EnsRegistry = await deploy('ENSRegistry')
+
+    BaseRegistrar = await deploy(
+      'BaseRegistrarImplementation',
+      EnsRegistry.address,
+      namehash('eth')
+    )
+
+    await BaseRegistrar.addController(account)
+    await BaseRegistrar.addController(account2)
+
+    MetaDataservice = await deploy(
+      'StaticMetadataService',
+      'https://ens.domains'
+    )
+
+    NameWrapper = await deploy(
+      'NameWrapper',
+      EnsRegistry.address,
+      BaseRegistrar.address,
+      MetaDataservice.address
+    )
+    PublicResolver = await deploy(
+      'PublicResolver',
+      EnsRegistry.address,
+      NameWrapper.address,
+      EMPTY_ADDRESS,
+      EMPTY_ADDRESS
+    )
+
+    // setup .eth
+    await EnsRegistry.setSubnodeOwner(
+      ROOT_NODE,
+      utils.keccak256(utils.toUtf8Bytes('eth')),
+      BaseRegistrar.address
+    )
+
+    // setup .xyz
+    await EnsRegistry.setSubnodeOwner(
+      ROOT_NODE,
+      utils.keccak256(utils.toUtf8Bytes('xyz')),
+      account
+    )
+
+    //make sure base registrar is owner of eth TLD
+    expect(await EnsRegistry.owner(namehash('eth'))).to.equal(
+      BaseRegistrar.address
+    )
+
+    SubdomainRegistrar = await deploy('SubdomainRegistrar', NameWrapper.address)
+
+    SubdomainRegistrar2 = SubdomainRegistrar.connect(signers[1])
+  })
+
+  beforeEach(async () => {
+    result = await ethers.provider.send('evm_snapshot')
+  })
+  afterEach(async () => {
+    await ethers.provider.send('evm_revert', [result])
+  })
+
+  describe('register', () => {
+    it('should allow subdomains to be created', async () => {
+      await BaseRegistrar.register(labelhash('test'), account, 86400)
+      await BaseRegistrar.setApprovalForAll(NameWrapper.address, true)
+      await NameWrapper.wrapETH2LD(
+        'test',
+        account,
+        0,
+        MAX_EXPIRY,
+        EMPTY_ADDRESS
+      )
+      expect(await NameWrapper.ownerOf(node)).to.equal(account)
+      await NameWrapper.setApprovalForAll(SubdomainRegistrar.address, true)
+      await SubdomainRegistrar2.register(
+        node,
+        'subname',
+        account2,
+        EMPTY_ADDRESS,
+        0,
+        86400,
+        []
+      )
+
+      expect(await NameWrapper.ownerOf(subNode)).to.equal(account2)
+    })
+    it('should not allow unapproved subdomains to be created', async () => {
+      await BaseRegistrar.register(labelhash('test'), account, 86400)
+      await BaseRegistrar.setApprovalForAll(NameWrapper.address, true)
+      await NameWrapper.wrapETH2LD('test', account, 0, 0, EMPTY_ADDRESS)
+      expect(await NameWrapper.ownerOf(namehash('test.eth'))).to.equal(account)
+
+      await expect(
+        SubdomainRegistrar.register(
+          namehash('test.eth'),
+          'subname',
+          account2,
+          EMPTY_ADDRESS,
+          0,
+          86400,
+          []
+        )
+      ).to.be.revertedWith(
```

Comment:
FYI, if this [PR](https://github.com/ensdomains/ens-contracts/pull/99) gets merged, this test will start failing as I enabled `throwOnCallFailures: false` to be able to test CCIP-read.

Please refer to [this part](https://github.com/ensdomains/ens-contracts/pull/99/files#diff-53ada1be0b3eb0cde6a37d29a3855a41bd5061aba9871e3991d1c0dbe896ee78R152) for the fix